### ISSUE TITLE
browse & thumbnail queueing: Add upcid to queue entries

### ISSUE THREAD

**Issue by dpmayerUSGS** [USGS-Astrogeology/PDS-Pipelines#457]
Currently, the browse/thumbnail queueing scripts create Redis queue entries containing  a path to a file, a fileid, and the archive clean name. For example:
`"('/pds_san/PDS_Archive/Galileo/SSI/go_0022/i24/io/c052079/6178r.lbl', 824535, 'galileo_ssi_edr')"`

It would be helpful for derived_process.py (or whatever the merged browse/thumbnail process script gets named in #453) if the Redis queue elements it consumes includes the upcid for a given product. This will enable the processing script to simply query the UPC database for the record in the json_keywords table for that product, and thus know which JSON blob to append the URL of the thumbnail/browse image to.

I'm not actually sure from examining the queueing scripts what `fileid` is, but I do know that it's *not* the `upcid` value associated with a given product. I think `fileid` can come out of the queue entries and be replaced with `upcid`.

**Updated to Add**
I've re-labeled this as a bug because it is valid for items in the UPC database to need browse/thumbnails created, but *not* have an ISIS serial number. This will always be the case with Level-2 products that are UPC-eligible, such as HiRISE RDRs. Therefore, we need a method of associating a given file with its UPC database entry that doesn't rely on the presence if an ISIS serial number. I think that using `upcid` would make sense, but I recognize based on Austin's comment below that implementing this might require a change to the UPC schema.

**Comment by AustinSanders** [comment]
Browse and thumbnail search based on archive / volume to find all matching products in the Files table.  Files contains only the file id, and not the upc id, which is contained in DataFiles, SearchTerms, and JsonKeywords.  Interestingly, none of these tables share foreign keys with Files.

This means that, in order to grab the upcid, we have to calculate and search for the ISIS id in the DataFiles table.  This functionality is currently in derived_process.  If we're just moving the functionality to a new spot, I'm wondering if it's worth pursuing.

My initial thought is that it is not.  If there is no change in efficiency, then it seems like the functionality should just exist where it best fits.  In a producer / consumer relationship like this, it makes the most sense (to me) to produce some minimal, generic set of resources and allow consumers to use / modify them on a per-need basis.  That being said, the expected one-to-one coupling of this particular producer/consumer relationship seems to de-emphasize the need for a true interface specification.  

TLDR:
upcid isn't available in browse / thumbnail queueing, so we would have to look it up the same way we're doing in the processing script.  I can't think of a good reason to move the logic to queueing since it is no more computationally efficient and would increase data transfer volume.

**Comment by AustinSanders** [comment]
I've thought more about this, and I'm beginning to wonder why there is a distinction between fileid and upcid.  We certainly seem to use upcid more, but I'm actually going to argue against its existence.

Files are originally ingested into the FIles table, at which point they are each assigned a unique fileid.  A subset of these files are marked as upc eligible and are processed / stored in the DataFiles table, at which point they are assigned a NEW, unique id that only exists for upc eligible items.

But.... why? Why not just use the existing fileid?  The files in the UPC database *are* the same files as those in the Files table, which is what allows us to look up the files based on ISIS serial number.  We could even still call the field upcid in order to minimize refactors.  The only difference here is that instead of setting the upcid column as an auto-incrementing integer, we could just say upcid=fileid.  This would allow for a direct lookup based on fileid rather than performing a serial-number-based search.  It also solidifies the existing (but currently implicit) relationship between records in Files and in DataFiles.

The infrastructure for this change is already in place on the Python side -- we're passing the fileid around via queues.  This method would be minimally impactful on the schema, and it would improve the time complexity of the current system by cutting out the ISIS SN calculation and string-based lookup.

I'll continue thinking about considerations / drawbacks and update the thread accordingly.

**Comment by USGSsakins** [comment]
Even if we do not use the UPCID, it is necessary for the database to have a unique ID assigned by the database for efficient searches and joining the tables together.  We do not need to ever expose this ID, but we do need it to ensure we can query all of the data in the schema.

**Comment by AustinSanders** [comment]
@USGSsakins Interesting.  To make sure that I understand correctly, there is a performance difference in a unique ID generated by the database and a unique ID that is specified on record entry?

**Comment by USGSsakins** [comment]
@AustinSanders As far as I understand it, the ID's generated by the DB are tied to the OID (object ID) that the database uses internally to track rows.  I've never tried creating my own ID's independent of the ones from a database sequence to compare search response speeds.

**Comment by AustinSanders** [comment]
Hmm.  That's a bummer.  I might play around a bit and try to get some benchmarks.  Thanks for the info!

**Comment by dpmayerUSGS** [comment]
Just updating to clarify something that we missed initially: fileid is the numerical identifier from the *DI database*. The only reason that it is included when queueing files for UPC or Derived processing is in order to create an entry of process_runs table of the DI database. 

I'm not sure there's much point to the process_runs table anymore as, from a user standpoint, we're not using it for anything. I'll open a separate issue about that.

If we decide to stop bothering to write anything to the process_runs table, then `derived_queueing.py` could simply query the UPC database based on matching string patterns in the DataFiles table in order to identify the file (and associated upcid) to create browse/thumbnail images for.

I opened #512 to deal with the narrowly-defined issue of ISIS serial numbers not being required or even unique.
